modules/index/src/main/java/org/apache/ignite/internal/idx/IndexManagerImpl.java

```diff
@@ -0,0 +1,515 @@
+        if (!busyLock.enterBusy()) {
+            throw new IgniteException(new NodeStoppingException());
+        }
+        try {
+            return indexAsyncInternal(idxCanonicalName);
+        } finally {
+            busyLock.leaveBusy();
+        }
+    }
+
+    private CompletableFuture<InternalSortedIndex> indexAsyncInternal(String idxCanonicalName) {
+        var idx = idxsByName.get(idxCanonicalName);
+
+        if (idx != null) {
+            return CompletableFuture.completedFuture(idx);
+        }
+
+        CompletableFuture<InternalSortedIndex> getIdxFut = new CompletableFuture<>();
+
+        EventListener<IndexEventParameters> clo = new EventListener<>() {
+            @Override
+            public boolean notify(@NotNull IndexEventParameters parameters, @Nullable Throwable e) {
+                if (!idxCanonicalName.equals(parameters.indexName())) {
+                    return false;
+                }
+
+                if (e == null) {
+                    getIdxFut.complete(parameters.index());
+                } else {
+                    getIdxFut.completeExceptionally(e);
+                }
+
+                return true;
+            }
+
+            @Override
+            public void remove(@NotNull Throwable e) {
+                getIdxFut.completeExceptionally(e);
+            }
+        };
+
+        listen(IndexEvent.CREATE, clo);
+
+        idx = idxsByName.get(idxCanonicalName);
+
+        if (idx != null && getIdxFut.complete(idx) || !isIndexConfigured(idxCanonicalName) && getIdxFut.complete(null)) {
+            removeListener(IndexEvent.CREATE, clo, null);
+        } else {
+            getIdxFut.thenRun(() -> removeListener(IndexEvent.CREATE, clo, null));
+        }
+
+        return getIdxFut;
+    }
+
+    /**
+     * Checks that the index is configured with specific name.
+     *
+     * @param idxName Index name.
+     * @return True when the table is configured into cluster, false otherwise.
+     */
+    private boolean isIndexConfigured(String idxName) {
+        NamedListView<TableView> directTablesCfg = directProxy(tablesCfg.tables()).value();
+
+        // TODO: IGNITE-15721 Need to review this approach after the ticket would be fixed.
+        // Probably, it won't be required getting configuration of all tables from Metastore.
+        for (String name : directTablesCfg.namedListKeys()) {
+            TableView tableView = directTablesCfg.get(name);
+
+            if (tableView != null) {
+                if (tableView.indices().get(idxName) != null) {
+                    return true;
+                }
+            }
+        }
+
+        return false;
+    }
+
+
+    /**
+     * Waits for future result and return, or unwraps {@link CompletionException} to {@link IgniteException} if failed.
+     *
+     * @param future Completable future.
+     * @return Future result.
+     */
+    private <T> T join(CompletableFuture<T> future) {
+        if (!busyLock.enterBusy()) {
```

Comment:
This will prevent node from stopping is the one called sync operation.
However, the behavior will be different when waiting on future that were got from async method.

Reply:
Dropped with sync internal API